src/checkpoint.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright 2018 Eugene P.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+
+use {DB, Error};
+use ffi;
+use std::ffi::CString;
+use std::path::Path;
+
+/// Undocumented parameter for `ffi::rocksdb_checkpoint_create` function. Zero by default.
+const LOG_SIZE_FOR_FLUSH: u64 = 0_u64;
+
+/// Database's checkpoint object.
+/// Used to create checkpoints of the specified DB from time to time.
+pub struct CheckpointBuilder {
+    inner: *mut ffi::rocksdb_checkpoint_t,
+}
+
+impl CheckpointBuilder {
```

Comment:
I'd recommend this is changed back to `Checkpoint`. That is what the underlying object is named in the [core RocksDB library](https://github.com/facebook/rocksdb/blob/master/include/rocksdb/utilities/checkpoint.h), and is similarly named in the [Java](https://github.com/facebook/rocksdb/blob/master/java/src/main/java/org/rocksdb/Checkpoint.java) and [Go](https://github.com/tecbot/gorocksdb/blob/master/checkpoint.go) language bindings.

Reply:
This was my initial thought on how to name it. 

@vitvakatu should I rename it back?

Reply:
I can agree with you, the consistency of code base among different bindings and original project is important, but don't you think that `Checkpoint` naming is misleading? I can't understand why `Checkpoint` does not have `create` method (perhaps static?), but has verbose `CreateCheckpoint`. One of the reason may be consistency with C API, but again, that looks weird to me.

Reply:
To be more clear, I'd rename:
* `CheckpointBuilder` to `Checkpoint`
* `create_checkpoint` to `save_to`

So the workflow becomes:

```rust
let checkpoint = Checkpoint::new(); // Internal checkpoint_t object is created

let path = "cp1";
checkpoint.save_to(&cp1)?; // Physical directory produced
```

Reply:
@eupn I like your idea, yep. Of course, every review comment is my personal opinion and should be considered :wink:

Reply:
Thanks all - good discussion here.

I agree, the underlying API is not ideal.  I can't even discern why there even needs to be a `Checkpoint` class/object in the core RocksDB library - it seems like it would have been simpler to just put a `create_checkpoint` method right on the `DB` object.

That said, I'd like to make an argument to leave the function name as `create_checkpoint` instead of changing it to `save_to`.  Even though we all seem to agree that `save_to` is a much better name, it now creates a mismatch between this crate and underlying core library.  Each time we change the names of things it requires more work for the users of the crate to translate between the two.  The [official documentation](https://github.com/facebook/rocksdb/wiki/Checkpoints) on Checkpoints states that you first create the `Checkpoint` object, and then call the `CreateCheckpoint` method.  If we keep the name `create_checkpoint`, the official documentation works very well for this crate.  If we change the name, the user will likely have to do more work to verify that `save_to` is just `create_checkpoint` with a different name.

I recall reading somewhere - and just spent the last 30 minutes trying to find the reference, unsuccessfully - that when creating a Rust crate that is wrapping an underlying library, the suggested course of action is to keep the wrapper crate thin and as 1-to-1 as possible.  The idea being that developers who have used the underlying library in other languages will not experience an "impedance mismatch", so to speak, when using the wrapper crate.  It also noted that if there is a strong desire to create an "opinionated" Rust create that does drastically change the name and shape of things, it suggested that two crates be created - one that is thin and 1-to-1, and one that is opinionated.  I wish I could find that reference - if I do, I will post it.  To be clear, I'm not suggesting we create another crate.

Since the official documentation, the [C API](https://github.com/facebook/rocksdb/blob/0bd2ede10e97944798449e0f954658fac8bfdbd2/include/rocksdb/c.h#L200-L202), and the [Java](https://github.com/facebook/rocksdb/blob/master/java/src/main/java/org/rocksdb/Checkpoint.java#L49-L52) and [Go](https://github.com/tecbot/gorocksdb/blob/master/checkpoint.go#L36-L50) language bindings all use idiomatic versions of `create_checkpoint`, as much as I think we all agree it's not the greatest API, I suggest we keep the `create_checkpoint` function name.

Reply:
@iSynaptic renamed it back

Reply:
@eupn @vitvakatu Thanks!